rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+  in (3.3), re-using an index allows us to detect these problems before trying
+  and avoid errors like (3.3.1) altogether.
+- Single index to backup instead of “index pattern” that matches any
+  `.kibana_n`.
+- Simplifies Kibana system index Elasticsearch plugin since it needs to work
+  on one index per "tenant".
+- Can prevent data loss when mixed versions of Kibana are both attempting a
+  migration using optimistic concurrency control.
+
+Drawbacks:
+- Cannot make breaking mapping changes (even though it was possible, we have not
+  introduced a breaking mapping change during 7.x).
+- Rollback is only possible by restoring a snapshot which requires educating
+  users to ensure that they don't rely on `.kibana_n` indices as backups.
+  (Apart from the need to educate users, snapshot restores provide many
+  benefits).
+- It narrows the second restriction under (4.3.1) even further: migrations
+  cannot rely on any state that could change as part of a migration because we
+  can no longer use the previous index as a snapshot of unmigrated state.
+- We can’t automatically perform a rollback from a half-way done migration.
+- It’s impossible to provide read-only functionality for outdated nodes which
+  means we can't achieve goal (2.7).
+
+### 4.5.2 Minimizing data loss with mixed Kibana versions (8.0)
+When multiple versions of Kibana are running at the same time, writes from the
+outdated node can result in outdated documents being written to the Kibana
+index. Since these outdated documents might have different fields, queries
+might only return a subset of the results which leads to incorrect results and
+even data loss.
+
+To prevent data loss we need to implement an index-wide “guard” which will
+prevent any writes from outdated nodes. The index-wide building blocks which
+Elasticsearch provides such as aliases or read-only indices are located inside
+Elasticsearch's ClusterState. But, there are certain failure scenarios under
+which ClusterState could remain inconsistent between nodes for an unbounded
+amount of time. In addition, any in-progress writes such as a bulk operation
+will continue to run even after an index was set to read-only or an alias had
+been removed.
+
+Since Elasticsearch doesn't expose the required building blocks, Kibana cannot
+guarantee that there would be no data loss. Instead, we aim to minimize it as
+much as possible.
+
+To perform a rollback for a failed migration, any "guards" put in place by
+newer Kibana nodes will have to be removed. By locating this state in the
+backup snapshot taken before an upgrade, rolling back to the previous snapshot
+will automatically remove the guard.
+
+1. Disable `action.auto_create_index` for the Kibana system indices.
+2. If a node detects that the `.kibana` index requires a migration it does the
+   following:
+   1. Create a version-specific read and write alias for sending commands
+       (.e.g `.kibana_8.0.1_read` & `.kibana_8.0.1_write`.
```

Comment:
If migrations are run when plugins are installed/removed, does this affect our use of version-specific aliases?

Reply:
These version aliases prevent (minimize) the chance of incompatible Kibana versions writing to the index and causing data loss. Two kibana instances on the same version but with different plugins are always compatible because they will write different types. So instance `a` with `pluginA` enabled will write `plugin_a_type` saved objects, but even though instance `b` doesn't know how to migrate the `plugin_a_type` because `pluginA` is disabled, it also will never read / write these objects.

Reply:
I should collapse these separate read/write alias into a single version-specific alias because we can't provide read-only functionality with in-place migrations so having two aliases serves no purpose.

Reply:
We only use a single version-specific alias for read/writes now.